### PR TITLE
HPCC-18006 Couchbase Plugin: Child datasets w/o XPATH

### PR DESCRIPTION
**Pull request by dcamper**
Fix issue with parsing results containing child datasets when those children did not have explicit XPATHs defined for them.

Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Against sample code provided in Jira ticket and with other well-known Couchbase data.

### CODE REVIEW

**Comment by dcamper** [comment]
@rpastrana Please review

File: plugins/couchbase/couchbaseembed.cpp [hunk 1]
```diff
@@ -1031,66 +1037,82 @@ namespace couchbaseembed
...
-                xpath = field->xpath;
+                const char * sep = strchr(field->xpath, xpathCompoundSeparatorChar);
+
+                if (!sep)
+                {
+                    outXPath.append(field->xpath);
+                }
+                else
+                {
+                    outXPath.append(field->xpath, 0, static_cast<size32_t>(sep - field->xpath));
```
Comment (rpastrana) [inline plugins/couchbase/couchbaseembed.cpp]: is it possible to have more than one xpathCompoundSeparatorChar in field->xpath? Does it matter?

**Review by rpastrana — COMMENTED** [review]
@dcamper looks good. I had one minor question. Also, can you remind me why the xpathorname is now taking a stringbuffer?

File: plugins/couchbase/couchbaseembed.cpp (hunk 1 above)
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.cpp]: It does not appear that xpathCompoundSeparatorChar will be present more than once. I found the construction for those values in ecl/hqlcpp/hqlhtcpp.cpp and the comment associated with it reads:

//Format of the xpath field is (nested-item 0x01 repeated-item)

Plus, a search through the codebase shows that all other code parsing for that character assumes that there is more than one present.

**Comment by dcamper** [comment]
CouchbaseRowBuilder::xpathOrName() used to return a simple const char* because all possible non-null values were already available. With this new version, one of the possible return values is a substring that requires construction.  It made sense to pass in a container for the constructed string at that point.

**Review by rpastrana — APPROVED** [review]
@dcamper looks good to go.

**Pull request merged**